linfa-trees/src/decision_trees/algorithm.rs

```diff
@@ -361,22 +489,36 @@ pub struct DecisionTree<F: Float, L: Label> {
     num_features: usize,
 }
 
-impl<F: Float, L: Label, D: Data<Elem = F>> Predict<ArrayBase<D, Ix2>, Vec<L>>
+impl<F: Float, L: Label, D: Data<Elem = F>> Predict<ArrayBase<D, Ix1>, L> for DecisionTree<F, L> {
+    /// Make prediction for sample `x`.
+    fn predict(&self, x: ArrayBase<D, Ix1>) -> L {
+        self.predict(&x)
+    }
+}
+
+impl<F: Float, L: Label, D: Data<Elem = F>> Predict<&ArrayBase<D, Ix1>, L> for DecisionTree<F, L> {
+    /// Make prediction for sample `x`.
+    fn predict(&self, x: &ArrayBase<D, Ix1>) -> L {
+        make_prediction(x, &self.root_node)
+    }
+}
+
+impl<F: Float, L: Label, D: Data<Elem = F>> Predict<ArrayBase<D, Ix2>, Array1<L>>
     for DecisionTree<F, L>
 {
     /// Make predictions for each row of a matrix of features `x`.
-    fn predict(&self, x: ArrayBase<D, Ix2>) -> Vec<L> {
-        x.genrows()
-            .into_iter()
-            .map(|row| make_prediction(&row, &self.root_node))
-            .collect()
+    fn predict(&self, x: ArrayBase<D, Ix2>) -> Array1<L> {
+        x.map_axis(Axis(1), |observation| {
+            make_prediction(&observation, &self.root_node)
+        })
```

Comment:
another implementation which came to my mind: take a `&Dataset<R, T>` and returns an `Array1<L>`, then you don't have to call `.predict(dataset.records())`, but just `.predict(&dataset)`

Reply:
Yes, that would really work well